Add a Docker Taskflow decorator

dimberman opened the pull request:
Add the ability to run @task.docker on a python function and turn it into a DockerOperator that can run that python function remotely.
```
@task.docker(
    image="quay.io/bitnami/python:3.8.8",
    force_pull=True,
    docker_url="unix://var/run/docker.sock",
    network_mode="bridge",
    api_version='auto',
)
def f():
    import random
    return [random.random() for i in range(10000000)]
```

One notable aspect of this architecture is that we had to build it to make as few assumptions about user setups as possible. We could not share a volume between the worker and the container as this would break if the user runs the airflow worker on a docker container. We could not assume that users would have any specialized system libraries on their images (this implementation only requires python 3 and bash).

To work with these requirements, we use base64 encoding to store a jinja generated python file and inputs (which are generated using the same functions used by the PythonVirtualEnvOperator). Once the container starts, it uses these environment variables to deserialize the strings, run the function, and store the result in a file located at /tmp/script.out.

Once the function completes, we create a sleep loop until the DockerOperator retrieves the result via docker's get_archive API. This result can then be deserialized using pickle and sent to Airflow's XCom library in the same fashion as a python or python_virtualenv result.


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

isaulv commented on tests/decorators/test_docker.py:
```diff
@@ -0,0 +1,170 @@
+        assert ti.xcom_pull() == {'number': test_number + 1, '43': 43}
+
+    def test_call_20(self):
+        """Test calling decorated function 21 times in a DAG"""
```
func name and doc string don't agree.

ashb commented on airflow/decorators/__init__.py:
```diff
@@ -139,5 +140,167 @@ def virtualenv(
+            cap_add=cap_add,
+            extra_hosts=extra_hosts,
+            **kwargs,
+        )
```
Couldn't this all just be:


```python
        docker = _docker_task
```

ashb commented on airflow/decorators/docker.py:
```diff
@@ -0,0 +1,159 @@
+from typing import Callable, Dict, Optional, TypeVar
+
+from airflow.decorators.base import DecoratedOperator, task_decorator_factory
+from airflow.providers.docker.operators.docker import DockerOperator
```
This import cant be at the top level -- it makes the docker provider a required dep of core as a result.

dimberman replied:
@ashb this might be more reason to make the decorator a part of the docker provider? We can then have the task object import as needed?

ashb replied:
We talked about making the "sub-decorators' (i.e. `@task.x`) pluggable.

ashb commented on airflow/decorators/docker.py:
```diff
@@ -0,0 +1,159 @@
+        f' x = base64.b64decode(os.environ[\\"{env_var}\\"]);'
+        f' f = open(\\"{file}\\", \\"wb\\"); f.write(x);'
+        f' f.close()"'
+    )
```
```suggestion
    return shlex.quote(
        f'python -c "import os; import base64;'
        f' x = base64.b64decode(os.environ["{env_var}"]);'
        f' f = open("{file}", "wb"); f.write(x);'
        f' f.close()"'
    )
```

dimberman replied:
I had to undo this change as it broke a bunch of tests and I couldn't figure out how to get it to work. Glad to discuss if you have ideas.

ashb commented on airflow/decorators/docker.py:
```diff
@@ -0,0 +1,159 @@
+def _b64_encode_file(filename):
+    data = open(filename, "rb").read()
+    encoded = base64.b64encode(data)
+    return encoded
```
"Leaking" a FH here (not strictly, as it'll be closed when GC'd)

```suggestion
    with open(filename, "rb") as fh:
        return  base64.b64encode(fh.read())
```

ashb commented on airflow/decorators/docker.py:
```diff
@@ -0,0 +1,159 @@
+                f'touch /tmp/string_args &&'
+                f'touch /tmp/script.in &&'
+                f'{_generate_decode_command("PYTHON_INPUT", "/tmp/script.in")} &&'
+                f'python /tmp/script.py /tmp/script.in /tmp/script.out /tmp/string_args \''
```
Why bash _and_ python -- why not just do this all in the one generated python command?

ashb commented on airflow/decorators/docker.py:
```diff
@@ -0,0 +1,159 @@
+
+    def _write_string_args(self, filename):
+        with open(filename, 'w') as file:
+            file.write('\n'.join(map(str, self.string_args)))
```
Two things here:

1. This is duplicating code from VenvOperator, we should fix a way to not need that (make a mixin perhaps?)
2. The whole `self.string_args` is a misfeature of Venv we shouldn't copy over anyway.

ashb commented on airflow/example_dags/tutorial_taskflow_api_etl_docker_virtualenv.py:
```diff
@@ -0,0 +1,120 @@
+        force_pull=True,
+        docker_url="unix://var/run/docker.sock",
+        network_mode="bridge",
+        api_version='auto',
```
```suggestion
```

Those are all the defaults, so we shouldn't need to include them in the example.

ashb reviewed (commented):
This needs some docs showing about it, including a giant warning about the downsides (that it ships _sourcecode_ by inspecting it, length, no closures etc.

I'm not sure this needs a _whole_ example dag (each one we add slows down tests, as all dags   get loaded by some dags.) At least we should put it in a separate location so it doesn't get loaded by anything that loads the full dag bags.

ashb commented on airflow/decorators/__init__.py:
```diff
@@ -139,5 +139,169 @@ def virtualenv(
 
+    @staticmethod
+    def docker(  # pylint: disable=too-many-arguments,too-many-locals
+        python_callable: Optional[Callable] = None,
```
Yeah, it would be _super_ good if we didn't have to duplicate the args, defaults and docstring from DockerOperator here.

dimberman replied:
I am all for that if we can do so in a way that still allows for IDE completion.

uranusjr replied:
I wonder if `functools.update_wrapper()` or `functools.wraps()` would work here.

dimberman committed:
Make decorators pluggable

This PR will allow users to add custom "@task.____" decorators by adding
to their setup.cfg files. This will make decorators seem native to
airflow while living in provider packages.

uranusjr commented on airflow/providers/docker/decorators/docker.py:
```diff
@@ -0,0 +1,159 @@
+    return shlex.quote(
+        f'python -c "import os; import base64;'
+        f' x = base64.b64decode(os.environ["{env_var}"]);'
+        f' f = open("{file}", "wb"); f.write(x);'
```
Theoratically `env_var` and `file` could contain the quote character (`"`), so it’s better to use `!r` instead.

uranusjr commented on airflow/decorators/__init__.py:
```diff
@@ -41,103 +45,12 @@ def __call__(
+            return self.store[name]
+        decorator = ProvidersManager().taskflow_decorators[name]
+        self.store[name] = decorator
+        return decorator
```
Using `__getattr__` like this breaks IDE introspection (e.g. autocomplete) and type hints so it’d be best to avoid it as much as possible. Maybe we could use metaclass or some kind of descriptor to achieve this.

dimberman commented on airflow/decorators/__init__.pyi:
```diff
@@ -0,0 +1,105 @@
+
+        except ImportError:
+            pass
+    def python(
```
@ashb and I spoke it over and for now we're just going to keep python and virtualenv in the taskdecorator class. It doesn't change anything functionally and unlike external providers, python and virtualenv will ONLY update with updates to Airflow.

potiuk replied:
Makes perfect sense!

kaxil commented on airflow/example_dags/tutorial_taskflow_api_etl_docker_virtualenv.py:
```diff
@@ -0,0 +1,117 @@
+    @task.virtualenv(
+        use_dill=True,
+        system_site_packages=False,
+        requirements=['funcsigs'],
```
Not strictly needed in this PR but a better example of adding a package here that is actually used in the function below would be nice and would be easier to understand from a use-case point of view too

ashb replied:
@dimberman Is `funcsigs` needed for this PR to work, or is this just meant to be an example?

kaxil commented on airflow/provider.yaml.schema.json:
```diff
@@ -210,7 +210,20 @@
+    },
+    "task-decorators": {
+        "type": "array",
+        "description": "Apply custom decorators to the TaskFlow API. Can be accessed by users via '@task.<name>'",
```
```suggestion
        "description": "Decorators to use with the TaskFlow API. Can be accessed by users via '@task.<name>'",
```

ashb commented on airflow/providers/docker/decorators/docker.py:
```diff
@@ -0,0 +1,459 @@
+    """
+
+    template_fields = ('op_args', 'op_kwargs')
+    template_fields_renderers = {"op_args": "py", "op_kwargs": "py"}
```
Given this operates on a function, I don't think we want to render files here.

```suggestion
```

ashb commented on airflow/providers/docker/decorators/docker.py:
```diff
@@ -0,0 +1,459 @@
+        self.string_args = [1, 2, 1]
+        self._output_filename = ""
+        command = "dummy command"
+        self.pickling_library = pickle
```
Should support dill/anything else the same way VirtualEnv operator does please.

dimberman replied:
I'm going to do pickle and dill for now as those are the two options that our virtualenv operator supports. Not sure about what others there are.

ashb commented on airflow/providers/docker/decorators/docker.py:
```diff
@@ -0,0 +1,459 @@
+):
+    """
+    Python operator decorator. Wraps a function into an Airflow operator.
+    Accepts kwargs for operator kwarg. Can be reused in a single DAG.
```
```suggestion
    Also accepts any argument that DockerOperator will via ``kwargs``. Can be reused in a single DAG.
```

ashb commented on airflow/providers_manager.py:
```diff
@@ -278,6 +309,51 @@ def _get_attr(obj: Any, attr_name: str):
+        if decorator_name in self._taskflow_decorator_dict:
+            log.warning(
+                "The hook_class '%s' has been already registered.",
+                decorator_class_name,
```
```suggestion
                "The taskflow decoraotr '%s' has been already registered (by %s).",
                decorator_name, self._taskflow_decorator_dict[decorator_name].__name__
```

ashb commented on airflow/providers_manager.py:
```diff
@@ -278,6 +309,51 @@ def _get_attr(obj: Any, attr_name: str):
+        try:
+            module, class_name = decorator_class_name.rsplit('.', maxsplit=1)
+            decorator_class = getattr(importlib.import_module(module), class_name)
+            self._taskflow_decorator_dict[decorator_name] = decorator_class
```
This causes _all_ decorators to be imported up front, which was something I wanted us to avoid -- i.e. I'd like `airflow.providers.docker.decorators.docker.docker_decorator` to only be imported on the first access (per process) of `@task.docker`.

potiuk replied:
As discussed before - I will make another change/refactor to the "providers_manager" to accomplish that in "generic" way

dimberman replied:
@ashb so is this something that should be addressed in this PR or another? As it stands there won't be many (if any) decorators to import, so maybe we can defer?

dimberman replied:
@potiuk did you ever make a refactor for this? If not I'd like to figure out a way we can defer this issue so we can get this merged.

dimberman replied:
@ashb I think we can kick the can on this. For now there will be all of zero decorators that will be loaded in this case. I think we can reasonably assume that we're not going to see a flood of custom decorators in the near future. This can be a 2.2.1 feature etc.

ashb commented on docs/apache-airflow/docker_decorator.py:
```diff
@@ -0,0 +1,460 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
I don't think you meant to have this file here.

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
 
 This option should allow for far greater flexibility for users who wish to keep their workflows more simple
 and pythonic.
+Using the Taskflow API with Docker or Virtual Environments
```
```suggestion

Using the Taskflow API with Docker or Virtual Environments
```

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
+Using the Taskflow API with Docker or Virtual Environments
+----------------------------------------------------------
+
+As of Airflow <Airflow version>, you will have the ability to use the Taskflow API with either a
```
```suggestion
As of Airflow 2.2, you will have the ability to use the Taskflow API with either a
```

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
+----------------------------------------------------------
+
+As of Airflow <Airflow version>, you will have the ability to use the Taskflow API with either a
+docker container or python virtual environment. This added functionality will allow a much more
```
```suggestion
Docker container or Python virtual environment. This added functionality will allow a much more
```

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
+If you don't want to run your image on a docker environment, and instead want to create a separate virtual
+environment on the same machine, you can use the ``@task.virtualenv`` decorator instead. The ``@task.virtualenv``
+decorator will allow you to create a new virtualenv with custom libraries and even a different
+python version to run your function.
```
```suggestion
Python version to run your function.
```

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
+
+To use a docker image with the Taskflow API, change the decorator to ``@task.docker``
+and add any needed arguments to correctly run the task. Please note that the docker
+image must have a python library and take in a bash command as the ``command`` argument.
```
```suggestion
image must have a working Python installed and take in a bash command as the ``command`` argument.
```

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
+    :start-after: [START transform_docker]
+    :end-before: [END transform_docker]
+
+If you don't want to run your image on a docker environment, and instead want to create a separate virtual
```
```suggestion
If you don't want to run your image on a Docker environment, and instead want to create a separate virtual
```

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
+    :end-before: [END extract_virtualenv]
+
+These two options should allow for far greater flexibility for users who wish to keep their workflows more simple
+and pythonic.
```
Please add some info here about limitations you might need to place on input or output? (Size limits or output types?)

ashb commented on docs/apache-airflow/tutorial_taskflow_api.rst:
```diff
@@ -180,16 +180,155 @@ and even a different python version to run your function.
+These two options should allow for far greater flexibility for users who wish to keep their workflows more simple
+and pythonic.
+
+Creating Custom TaskFlow Decorators
```
This section should be in a separate doc in howto I hink

dimberman committed:
create docker decorator

dimberman committed:
Update docs/apache-airflow/tutorial_taskflow_api.rst

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

dimberman committed:
docs

dimberman committed:
Update airflow/provider.yaml.schema.json

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

dimberman committed:
split doc

dimberman committed:
added new file

dimberman committed:
try again doc

dimberman committed:
doc build passing

ashb commented on airflow/decorators/__init__.py:
```diff
@@ -15,34 +15,58 @@
+# [END import_docker]
+
 
 class _TaskDecorator:
```
We can mix these ones in directly:

```python
class _TaskDecorator(PythonDecoratorMixin, DockerDecoratorMixin):
```

dimberman replied:
@ashb I don't think you can import from None

```
Python 3.7.8 (default, Oct  8 2020, 09:28:20)
[Clang 12.0.0 (clang-1200.0.32.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class Foo(None):
...     def bar():
...             pass
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: NoneType takes no arguments
>>>
```

ashb replied:
Typo, I meant to say

```python
class _TaskDecorator(PythonDecoratorMixin, PythonVirtualenvDecoratorMixin)
```

ashb commented on airflow/decorators/__init__.py:
```diff
@@ -15,129 +15,58 @@
-
-from airflow.decorators.python import python_task
-from airflow.decorators.python_virtualenv import _virtualenv_task
+from airflow.decorators.python import PythonDecoratorMixin
```
```suggestion
from airflow.decorators.python import PythonDecoratorMixin, python_task
```

We probably have to keep this for back-compat reasons.

dimberman committed:
Update airflow/decorators/__init__.py

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

dimberman committed:
fix static

dimberman committed:
add docker init

dimberman committed:
fix tests

dimberman committed:
let's see if this fixes the import error

dimberman committed:
possibly fix docs

dimberman committed:
possibly fix docs

ashb commented on airflow/providers_manager.py:
```diff
@@ -531,6 +555,52 @@ def _get_attr(obj: Any, attr_name: str):
+                return
+        if decorator_name in self._taskflow_decorator_dict:
+            log.warning(
+                "The taskflow decoraotr '%s' has been already registered (by %s).",
```
```suggestion
                "The taskflow decorator '%s' has been already registered (by %s).",
```

ashb commented on docs/apache-airflow/howto/create-custom-decorator.rst:
```diff
@@ -0,0 +1,118 @@
+2. Create a ``foo_task`` function
+
+Once you have your decorated class, create a function that takes arguments ``python_callable``\, ``multiple_outputs``\,
+and ``kwargs``\. This function will use the ``airflow.decorators.base.task_decorator_factory`` function to convert
```
```suggestion
Once you have your decorated class, create a function that takes arguments ``python_callable``, ``multiple_outputs``,
and ``kwargs``. This function will use the ``airflow.decorators.base.task_decorator_factory`` function to convert
```

I don't think the `\` does anything here, does it?

ashb commented on docs/apache-airflow/howto/create-custom-decorator.rst:
```diff
@@ -0,0 +1,118 @@
+
+Finally, add a key-value of ``decorator-name``:``path-to-function`` to your provider.yaml. When Airflow starts, the
+``ProviderManager`` class will automatically import this value and ``task.decorator-name`` will work as a new
+decorator!
```
Please mention that the key here must be a valid python identifier (and change your example so it is) -- cos `@task.decorator-name` isn't going to work 🙂

ashb commented on docs/apache-airflow/howto/create-custom-decorator.rst:
```diff
@@ -0,0 +1,118 @@
+       `Docker <https://docs.docker.com/install/>`__
+
+   task-decorators:
+       docker: airflow.providers.docker.operators.docker.docker_decorator
```
Use the same "Foo" example here, not docker for consistency

ashb commented on docs/apache-airflow/howto/create-custom-decorator.rst:
```diff
@@ -0,0 +1,118 @@
+through stub files, but we wanted to avoid any situation where a user would update their provider and the autocomplete
+would be out of sync with the provider's actual parameters.
+
+To hack around this problem, we found that you could extend the ``_TaskDecorator`` class in the ``__init__.pyi`` file
```
We don't do it in the `pyi` file anymore.

ashb commented on docs/apache-airflow/howto/create-custom-decorator.rst:
```diff
@@ -0,0 +1,118 @@
+.. exampleinclude:: ../howto/docker_decorator.py
+    :language: python
+    :start-after: [START decoratormixin]
+    :end-before: [END decoratormixin]
```
Rather than including a whole copy of docker_decorator.py here (that will get out of date) can we instead point at the actual airflow/providers/docker/decorators/docker.py?

ashb committed:
Tidy up decorators

ashb reviewed (dismissed):
tests/decorators/test_docker.py should be under tests/provider/docker/ somewhere.

Please use pytest fixtures instead of a unittest.TestCase+setup/teardown for tests.

We still talk about the .pyi in the docs, but the implementation uses mixin for Docker directly in the .py --- shouldn't we register it via provider manager even for the docker provider?

ashb committed:
Review feedback

mik-laj commented on docs/spelling_wordlist.txt:
```diff
@@ -141,6 +141,7 @@ Firehose
 Firestore
 Flink
 FluentD
+FooDecoratedOperator
```
```suggestion
```
code-block are excluded from spell checking.

ashb committed:
fixup! Review feedback

kaxil commented on airflow/decorators/__init__.py:
```diff
@@ -15,129 +15,36 @@
+class _TaskDecorator(PythonDecoratorMixin, PythonVirtualenvDecoratorMixin):
+    def __getattr__(self, name):
+        if name.startswith("__"):
+            raise AttributeError(f'{type(self).__name__} has not attriubte {name!r}')
```
```suggestion
            raise AttributeError(f'{type(self).__name__} has no attribute {name!r}')
```

ashb replied:
What a mishmash of typing there. Shesh Ash 🙂

ashb committed:
fixup! Review feedback

kaxil commented on airflow/providers/docker/decorators/docker.py:
```diff
@@ -0,0 +1,298 @@
+        use_dill=False,
+        **kwargs,
+    ) -> None:
+        self.string_args = [1, 2, 1]
```
What's this?

kaxil replied:
Why `[1, 2, 1]`

ashb replied:
Vestigial code. Removed

ashb committed:
fixup! Review feedback

ashb committed:
fixup! Review feedback
